gdal/frmts/netcdf/netcdflayersg.cpp

```diff
@@ -0,0 +1,194 @@
+/* NetCDF Simple Geometry specific implementations
+ * for layers filled from CF 1.8 simple geometry variables
+ *
+ */
+#include "netcdfsg.h"
+#include "netcdfdataset.h"
+#include "ogr_core.h"
+
+namespace nccfdriver
+{
+	static OGRwkbGeometryType RawToOGR(geom_t type)
+	{
+		OGRwkbGeometryType ret = wkbNone;
+
+		switch(type)
+		{
+			case NONE:
+				break;
+			case LINE:
+				ret = wkbLineString;
+				break;
+			case MULTILINE:
+				ret = wkbMultiLineString;
+				break;
+			case POLYGON:
+				ret = wkbPolygon;
+				break;
+			case MULTIPOLYGON:
+				ret = wkbMultiPolygon;
+				break;
+			case POINT:
+				ret = wkbPoint;
+				break;
+			case MULTIPOINT:
+				ret = wkbMultiPoint;
+				break;
+			case UNSUPPORTED:
+				break;
+		}
+
+		return ret;	
+	}	
+
+}
+
+CPLErr netCDFDataset::DetectAndFillSGLayers(int ncid)
+{
+	// Discover simple geometry variables
+	int var_count;
+	nc_inq_nvars(ncid, &var_count);	
+	std::vector<int> vidList;
+
+	nccfdriver::scanForGeometryContainers(ncid, vidList);	
+
+	// To still do: support multiple geometry containers
+
+	for(size_t itr = 0; itr < vidList.size(); itr++)
+	{
+		try
+		{
+			LoadSGVarIntoLayer(ncid, vidList[itr]);
+
+		}
+		
+		catch(nccfdriver::SG_Exception * e)
+		{
+			CPLError(CE_Warning, CPLE_AppDefined, "Translation of a simple geometry layer has been terminated prematurely due to an error.");
+			CPLError(CE_Warning, CPLE_AppDefined, "%s", e->get_err_msg());
+			delete e;
+		}
+	}
+
+	return CE_None;
+}
+
+CPLErr netCDFDataset::LoadSGVarIntoLayer(int ncid, int nc_basevarId)
+{
+	nccfdriver::SGeometry * sg = new nccfdriver::SGeometry(ncid, nc_basevarId);
+	int cont_id = sg->getContainerId();
+	nccfdriver::SGeometry_PropertyScanner pr(ncid, cont_id);
+	OGRwkbGeometryType owgt = nccfdriver::RawToOGR(sg->getGeometryType());
+
+	// Geometry Type invalid, avoid further processing
+	if(owgt == wkbNone)
+	{
+		delete sg;
+		throw new nccfdriver::SG_Exception_BadFeature();
+	}
+	
+
+	char baseName[NC_MAX_CHAR + 1];
+	memset(baseName, 0, NC_MAX_CHAR + 1);
+	nc_inq_varname(ncid, nc_basevarId, baseName);
+
+	netCDFLayer * poL = new netCDFLayer(this, ncid, baseName, owgt, nullptr); 
+
+	poL->EnableSGBypass();
+	OGRFeatureDefn * defn = poL->GetLayerDefn();
+	defn->SetGeomType(owgt);
+
+	size_t shape_count = sg->get_geometry_count();
+
+	// Add properties
+	std::vector<int> props = pr.ids();
+	for(size_t itr = 0; itr < props.size(); itr++)
+	{
+		poL->AddField(props[itr]);	
+	}
+
+	for(size_t featCt = 0; featCt < shape_count; featCt++)
+	{
+		OGRGeometry * featureDesc;
```

Comment:
ditto :+1:   will fix

Reply:
okay so I tried using a unique_ptr here and I got a seg. fault. GDB shows that the seg. fault occurs in the translation phase. Besides, according to the docs, ownership is handed off to the SetGeometryDirectly function so I'm going to for now keep it as a standard unmanaged pointer OGRGeometry*.

Reply:
you must do SetGeometryDirectly(the_geom.release()) to transfer ownership

Reply:
Oh okay interesting thanks, I'll take another look then later :)

Reply:
On the note of `unique_ptr` for the OGRGeometry * geometry, netCDFLayer * poL:

I took a look into `unique_ptr` looking through the [MSDN Standard Library Reference](https://docs.microsoft.com/en-us/cpp/standard-library/unique-ptr-class?view=vs-2019), which says that the object referred to by the `unique_ptr` is destroyed when the `unique_ptr` itself is destroyed (in the same manner as the legacy `auto_ptr` which I'm more familiar with). Unfortunately for a auto store / stack allocated `unique_ptr` as what is described here for poL and geometry, their `unique_ptr` s will be destroyed at the end of the function (on frame deconstruction), destroying the respective OGRGeometry or poL with them at the same time.

This being the case, I have opted to utilize OGR interface directly by transfering ownership, since sending something managed throughout the rest of OGR would probably require refactoring upstream into OGRGeometry and OGRLayer, probably by switching around or adding such interfaces using `unique_ptr`s or probably more fittingly using the `shared_ptr`. Other workarounds can be done, but as far as I know they will probably require explicit unmanaged heap usage anyway (defeating the purpose of a managed pointer in the first place).